### ISSUE TITLE
rsyslogd 1904 hung for a while and not easy to reproduce

### ISSUE THREAD

**Issue by fduan001** [rsyslog/rsyslog#4250]
### Steps to reproduce the behavior
Hard to reproduce
### Environment
- rsyslog version: 1904
- platform: Yocto
- for configuration questions/issues, include rsyslog.conf and included config files
`# rsyslog configuration file (for Red Hat-based systems)
# note that most of this config file uses old-style format,
# because it is well-known AND quite suitable for simple cases
# like we have with the default config. For more advanced
# things, RainerScript configuration is suggested.
#
# For more information see /usr/share/doc/rsyslog-*/rsyslog_conf.html
# or latest version online at http://www.rsyslog.com/doc/rsyslog_conf.html
# If you experience problems, see http://www.rsyslog.com/doc/troubleshoot.html

# avoid-startup-deadlock.conf
# This configuration is a workaround for 
# https://github.com/rsyslog/rsyslog/issues/1188
global(processInternalMessages="on")

#### MODULES ####

# EXA-5390 - ratelimit appears to cause memory leaks. Disabled it. 12/02/2014
#module(load="imuxsock" SysSock.RateLimit.Interval="2" SysSock.RateLimit.Burst="200" SysSock.RateLimit.Severity="5" SysSock.UsePIDFromSystem="on")
module(load="imuxsock" SysSock.UsePIDFromSystem="on")

module(load="imklog")   # provides kernel logging support (previously done by rklogd)
#module(load="immark")  # provides --MARK-- message capability
module(load="impstats" interval="600" severity="7")  # provides rsyslog statistics

#### GLOBAL DIRECTIVES ####

# Use default timestamp format
$ActionFileDefaultTemplate RSYSLOG_TraditionalFileFormat

# File syncing capability is disabled by default. This feature is usually not required,
# not useful and an extreme performance hit
#$ActionFileEnableSync on

# Suppress duplicate messages
$RepeatedMsgReduction on

#### RULES ####

# Log all kernel messages to the console.
# Logging much else clutters up the screen.
#kern.*                                                 /dev/console

# Log anything (except mail) of level info or higher.
# Don't log private authentication messages!
#*.info;mail.none;authpriv.none;cron.none                /var/log/messages

# The authpriv file has restricted access.
#authpriv.*                                              /var/log/secure

# Log all the mail messages in one place.
#mail.*                                                  /var/log/maillog

# Log cron stuff
#cron.*                                                  /var/log/cron

# Everybody gets emergency messages
#*.emerg                                                 :omusrmsg:*

# Save news errors of level crit and higher in a special file.
#uucp,news.crit                                          /var/log/spooler

# Save boot messages also to boot.log
#local7.*                                                /var/log/boot.log

### TEMPLATE ###
$IncludeConfig /etc/rsyslog.d/templates

##
##
##
##
## When maxsize is reached, it should call the "logrotate" with arg=/etc/logrotate.conf

$outchannel system_log,             /var/log/messages,            1000000, /usr/sbin/logrotate /etc/logrotate.conf
$outchannel important_log,          /mnt/pram1/logs/imessages,    1000000, /usr/sbin/logrotate /etc/logrotate.conf
$outchannel kernel_log,             /var/log/kernel,              1000000, /usr/sbin/logrotate /etc/logrotate.conf

$outchannel dot1xd_log,             /var/log/dot1xd.log,           500000, /usr/sbin/logrotate /etc/logrotate.conf
$outchannel aaad_log,               /var/log/aaad.log,             500000, /usr/sbin/logrotate /etc/logrotate.conf
$outchannel adjmgrd_log,            /var/log/adjmgrd.log,          500000, /usr/sbin/logrotate /etc/logrotate.conf
$outchannel aemgrd_log,             /var/log/aemgrd.log,           500000, /usr/sbin/logrotate /etc/logrotate.conf
$outchannel arcmgrd_log,            /var/log/arcmgrd.log,          500000, /usr/sbin/logrotate /etc/logrotate.conf
$outchannel audmgrd_log,            /var/log/audmgrd.log,          500000, /usr/sbin/logrotate /etc/logrotate.conf
$outchannel avonmgrd_log,           /var/log/avonmgrd.log,         500000, /usr/sbin/logrotate /etc/logrotate.conf
$outchannel bngaxd_log,             /var/log/bngaxd.log,           500000, /usr/sbin/logrotate /etc/logrotate.conf
$outchannel bngaxlid_log,           /var/log/bngaxlid.log,         500000, /usr/sbin/logrotate /etc/logrotate.conf
$outchannel bnglcmd_log,            /var/log/bnglcmd.log,          500000, /usr/sbin/logrotate /etc/logrotate.conf
#$outchannel bnglid_log,             /var/log/bnglid.log,           500000, /usr/sbin/logrotate /etc/logrotate.conf
$outchannel certmgrd_log,           /var/log/certmgrd.log,         500000, /usr/sbin/logrotate /etc/logrotate.conf
$outchannel chassismgrd_log,        /var/log/chassismgrd.log,      500000, /usr/sbin/logrotate /etc/logrotate.conf
$outchannel cmd_log,                /var/log/cmd.log,              500000, /usr/sbin/logrotate /etc/logrotate.conf
$outchannel confd_log,              /var/log/confd/confd.log,      500000, /usr/sbin/logrotate /etc/logrotate.conf
$outchannel cpeupgmgrd_log,         /var/log/cpeupgmgrd.log,       500000, /usr/sbin/logrotate /etc/logrotate.conf
$outchannel dadd_log,               /var/log/dadd.log,             500000, /usr/sbin/logrotate /etc/logrotate.conf
$outchannel dgtld_log,              /var/log/dgtld.log,            500000, /usr/sbin/logrotate /etc/logrotate.conf
$outchannel dhcpd_log,              /var/log/dhcpd.log,            500000, /usr/sbin/logrotate /etc/logrotate.conf
$outchannel dhclient_log,           /var/log/dhclient.log,         500000, /usr/sbin/logrotate /etc/logrotate.conf
$outchannel diagmgrd_log,           /var/log/diagmgrd.log,         500000, /usr/sbin/logrotate /etc/logrotate.conf
$outchannel diagtestd_log,          /var/log/diagtestd.log,        500000, /usr/sbin/logrotate /etc/logrotate.conf
$outchannel diameterd_log,          /var/log/diameterd.log,        500000, /usr/sbin/logrotate /etc/logrotate.conf
$outchannel diametergyd_log,        /var/log/diametergyd.log,      500000, /usr/sbin/logrotate /etc/logrotate.conf
$outchannel dmd_log,                /var/log/dmd.log,              500000, /usr/sbin/logrotate /etc/logrotate.conf
$outchannel dpmgrd_log,             /var/log/dpmgrd.log,           500000, /usr/sbin/logrotate /etc/logrotate.conf
$outchannel dyadmgrd_log,           /var/log/dyadmgrd.log,         500000, /usr/sbin/logrotate /etc/logrotate.conf
$outchannel eapsd_log,              /var/log/eapsd.log,            500000, /usr/sbin/logrotate /etc/logrotate.conf
$outchannel equipmentmgrd_log,      /var/log/equipmentmgrd.log,    500000, /usr/sbin/logrotate /etc/logrotate.conf
$outchannel erpsd_log,              /var/log/erpsd.log,            500000, /usr/sbin/logrotate /etc/logrotate.conf
$outchannel evtest_log,             /var/log/evtest.log,           500000, /usr/sbin/logrotate /etc/logrotate.conf
$outchannel fald_log,               /var/log/fald.log,             500000, /usr/sbin/logrotate /etc/logrotate.conf
$outchannel fccd_log,               /var/log/fccd.log,             500000, /usr/sbin/logrotate /etc/logrotate.conf
$outchannel fibd_log,               /var/log/fibd.log,            1100000, /usr/sbin/logrotate /etc/logrotate.conf
$outchannel flwmgrd_log,            /var/log/flwmgrd.log,          500000, /usr/sbin/logrotate /etc/logrotate.conf
$outchannel fprd_log,               /var/log/fprd.log,             500000, /usr/sbin/logrotate /etc/logrotate.conf
$outchannel gfastmgrd_log,          /var/log/gfastmgrd.log,        500000, /usr/sbin/logrotate /etc/logrotate.conf
$outchannel halm_bcmfastdsl_log,    /var/log/halm_bcmfastdsl.log,  500000, /usr/sbin/logrotate /etc/logrotate.conf
$outchannel halm_bcmxgsu_log,       /var/log/halm_bcmxgsu.log,     500000, /usr/sbin/logrotate /etc/logrotate.conf
$outchannel halm_xgs_log,           /var/log/halm_xgs.log,         820000, /usr/sbin/logrotate /etc/logrotate.conf
$outchannel halm_dyaddev_log,       /var/log/halm_dyaddev.log,     500000, /usr/sbin/logrotate /etc/logrotate.conf
$outchannel halm_robo_log,          /var/log/halm_robo.log,        500000, /usr/sbin/logrotate /etc/logrotate.conf
$outchannel halm_vmac_log,          /var/log/halm_vmac.log,        500000, /usr/sbin/logrotate /etc/logrotate.conf
$outchannel halm_vswitch_log,       /var/log/halm_vswitch.log,     500000, /usr/sbin/logrotate /etc/logrotate.conf
$outchannel halm_vitesse_log,       /var/log/halm_vitesse.log,     500000, /usr/sbin/logrotate /etc/logrotate.conf
$outchannel halm_np_log,            /var/log/halm_np.log,          500000, /usr/sbin/logrotate /etc/logrotate.conf
$outchannel halm_sckipio_log,       /var/log/halm_sckipio.log,     500000, /usr/sbin/logrotate /etc/logrotate.conf
$outchannel haltestd_log,           /var/log/haltestd.log,         500000, /usr/sbin/logrotate /etc/logrotate.conf
$outchannel halm_dnx_log,           /var/log/halm_dnx.log,         500000, /usr/sbin/logrotate /etc/logrotate.conf
$outchannel halm_sim_log,           /var/log/halm_sim.log,         820000, /usr/sbin/logrotate /etc/logrotate.conf
$outchannel halm_ovs_log,           /var/log/halm_ovs.log,         500000, /usr/sbin/logrotate /etc/logrotate.conf
$outchannel hatestd_log,            /var/log/hatestd.log,          500000, /usr/sbin/logrotate /etc/logrotate.conf
$outchannel gfastdslmgrd_log,       /var/log/gfastdslmgrd.log,     500000, /usr/sbin/logrotate /etc/logrotate.conf
$outchannel hostapd_log,            /var/log/hostapd.log,          500000, /usr/sbin/logrotate /etc/logrotate.conf
$outchannel hostmgrd_log,           /var/log/hostmgrd.log,         500000, /usr/sbin/logrotate /etc/logrotate.conf
$outchannel httprd_log,             /var/log/httprd.log            500000, /usr/sbin/logrotate /etc/logrotate.conf
$outchannel icld_log,               /var/log/icld.log,             500000, /usr/sbin/logrotate /etc/logrotate.conf
$outchannel ictpd_log,              /var/log/ictpd.log,            500000, /usr/sbin/logrotate /etc/logrotate.conf
$outchannel ifmgrd_log,             /var/log/ifmgrd.log,           500000, /usr/sbin/logrotate /etc/logrotate.conf
$outchannel ifmd_log,               /var/log/ifmd.log,             500000, /usr/sbin/logrotate /etc/logrotate.conf
$outchannel igmpd_log,              /var/log/igmpd.log,            500000, /usr/sbin/logrotate /etc/logrotate.conf
$outchannel ipmgrd_log,             /var/log/ipmgrd.log,           500000, /usr/sbin/logrotate /etc/logrotate.conf
$outchannel ipcmgrd_log,            /var/log/ipcmgrd.log,          500000, /usr/sbin/logrotate /etc/logrotate.conf
$outchannel ipdrd_log,              /var/log/ipdrd.log,            500000, /usr/sbin/logrotate /etc/logrotate.conf
$outchannel ipfixd_log,             /var/log/ipfixd.log,           500000, /usr/sbin/logrotate /etc/logrotate.conf
$outchannel l2dd_log,               /var/log/l2dd.log,             500000, /usr/sbin/logrotate /etc/logrotate.conf
$outchannel l2hostmgrd_log,         /var/log/l2hostmgrd.log,       500000, /usr/sbin/logrotate /etc/logrotate.conf
$outchannel l2mgrd_log,             /var/log/l2mgrd.log,           500000, /usr/sbin/logrotate /etc/logrotate.conf
$outchannel l2mmgrd_log,            /var/log/l2mmgrd.log,          500000, /usr/sbin/logrotate /etc/logrotate.conf
$outchannel lacpd_log,              /var/log/lacpd.log,            500000, /usr/sbin/logrotate /etc/logrotate.conf
$outchannel lldpd_log,              /var/log/lldpd.log,            500000, /usr/sbin/logrotate /etc/logrotate.conf
$outchannel lmd_log,                /var/log/lmd.log,             1100000, /usr/sbin/logrotate /etc/logrotate.conf
$outchannel loamd_log,              /var/log/loamd.log,            500000, /usr/sbin/logrotate /etc/logrotate.conf
$outchannel login_log,              /var/log/login.log,            500000, /usr/sbin/logrotate /etc/logrotate.conf
$outchannel logmgrd_log,            /var/log/logmgrd.log,          500000, /usr/sbin/logrotate /etc/logrotate.conf
$outchannel macsecd_log,            /var/log/macsecd.log,          500000, /usr/sbin/logrotate /etc/logrotate.conf
$outchannel notfmgrd_log,           /var/log/notfmgrd.log,         500000, /usr/sbin/logrotate /etc/logrotate.conf
$outchannel notfproxyd_log,         /var/log/notfproxyd.log,       500000, /usr/sbin/logrotate /etc/logrotate.conf
$outchannel ntpd_log,               /var/log/ntpd.log,             500000, /usr/sbin/logrotate /etc/logrotate.conf
$outchannel ofmgrd_log,             /var/log/ofmgrd.log,           500000, /usr/sbin/logrotate /etc/logrotate.conf
$outchannel olttcmgrd_log,          /var/log/olttcmgrd.log,        500000, /usr/sbin/logrotate /etc/logrotate.conf
$outchannel ontimgcd_log,           /var/log/ontimgcd.log,         500000, /usr/sbin/logrotate /etc/logrotate.conf
$outchannel ontimgfd_log,           /var/log/ontimgfd.log,         500000, /usr/sbin/logrotate /etc/logrotate.conf
$outchannel ontupgmd_log,           /var/log/ontupgmd.log,         500000, /usr/sbin/logrotate /etc/logrotate.conf
$outchannel ontupgsd_log,           /var/log/ontupgsd.log,         500000, /usr/sbin/logrotate /etc/logrotate.conf
$outchannel opentftpd_log,          /var/log/opentftpd.log,        500000, /usr/sbin/logrotate /etc/logrotate.conf
$outchannel pcmd_log,               /var/log/pcmd.log,             500000, /usr/sbin/logrotate /etc/logrotate.conf
$outchannel plmgrd_log,             /var/log/plmgrd.log,           500000, /usr/sbin/logrotate /etc/logrotate.conf
$outchannel ponmgrd_log,            /var/log/ponmgrd.log,          500000, /usr/sbin/logrotate /etc/logrotate.conf
$outchannel pppoeiad_log,           /var/log/pppoeiad.log,         500000, /usr/sbin/logrotate /etc/logrotate.conf
$outchannel qosmgrd_log,            /var/log/qosmgrd.log,          500000, /usr/sbin/logrotate /etc/logrotate.conf
$outchannel radiusd_log,            /var/log/radiusd.log,          500000, /usr/sbin/logrotate /etc/logrotate.conf
$outchannel rsyncd_log,             /var/log/rsyncd.log,           500000, /usr/sbin/logrotate /etc/logrotate.conf
$outchannel rstpd_log,              /var/log/rstpd.log,            500000, /usr/sbin/logrotate /etc/logrotate.conf
$outchannel rsyslogd_log,           /var/log/rsyslogd.log,         500000, /usr/sbin/logrotate /etc/logrotate.conf
$outchannel satmgrd_log,            /var/log/satmgrd.log,          500000, /usr/sbin/logrotate /etc/logrotate.conf
$outchannel servicemgrd_log,        /var/log/servicemgrd.log,      500000, /usr/sbin/logrotate /etc/logrotate.conf
$outchannel snmpd_log,              /var/log/snmpd.log,            500000, /usr/sbin/logrotate /etc/logrotate.conf
$outchannel soamd_log,              /var/log/soamd.log,            500000, /usr/sbin/logrotate /etc/logrotate.conf
$outchannel ssaliod_log,            /var/log/ssaliod.log,          500000, /usr/sbin/logrotate /etc/logrotate.conf
$outchannel ssalio2d_log,           /var/log/ssalio2d.log,         500000, /usr/sbin/logrotate /etc/logrotate.conf
$outchannel ssal_mgrd_log,          /var/log/ssal_mgrd.log,        500000, /usr/sbin/logrotate /etc/logrotate.conf
$outchannel sshd_log,               /var/log/sshd.log,             500000, /usr/sbin/logrotate /etc/logrotate.conf
$outchannel ssmd_log,               /var/log/ssmd.log,             500000, /usr/sbin/logrotate /etc/logrotate.conf
$outchannel tmmgrd_log,             /var/log/tmmgrd.log,           500000, /usr/sbin/logrotate /etc/logrotate.conf
$outchannel tracemgrd_log,          /var/log/tracemgrd.log,        500000, /usr/sbin/logrotate /etc/logrotate.conf
$outchannel uamgrd_log,             /var/log/uamgrd.log,           500000, /usr/sbin/logrotate /etc/logrotate.conf
$outchannel upgmgrd_log,            /var/log/upgmgrd.log,          500000, /usr/sbin/logrotate /etc/logrotate.conf
$outchannel usbmgrd_log,            /var/log/usbmgrd.log,          500000, /usr/sbin/logrotate /etc/logrotate.conf
$outchannel vcamgrd_log,            /var/log/vcamgrd.log,          500000, /usr/sbin/logrotate /etc/logrotate.conf
$outchannel vlanmgrd_log,           /var/log/vlanmgrd.log,         500000, /usr/sbin/logrotate /etc/logrotate.conf
$outchannel l2v6hostmgrd_log,       /var/log/l2v6hostmgrd.log,     500000, /usr/sbin/logrotate /etc/logrotate.conf
$outchannel salusd_log,             /var/log/salusd.log,           500000, /usr/sbin/logrotate /etc/logrotate.conf
$outchannel conversion_log,         /var/log/conversion.log,       500000, /usr/sbin/logrotate /etc/logrotate.conf
$outchannel pppsrvd_log,            /var/log/pppsrvd.log,          500000, /usr/sbin/logrotate /etc/logrotate.conf
$outchannel pppsrvdp_proxyd_log,    /var/log/pppsrvdp_proxyd.log,  500000, /usr/sbin/logrotate /etc/logrotate.conf
$outchannel vrrpd_log,              /var/log/vrrpd.log,            500000, /usr/sbin/logrotate /etc/logrotate.conf
$outchannel keepalived_log,         /var/log/keepalived.log,       500000, /usr/sbin/logrotate /etc/logrotate.conf

#### RULES ####
if $programname == 'dot1xd'             then :omfile:$dot1xd_log;exaLog
if $programname == 'aaad'               then :omfile:$aaad_log;exaLog
if $programname == 'adjmgrd'            then :omfile:$adjmgrd_log;exaLog
if $programname == 'aemgrd'             then :omfile:$aemgrd_log;exaLog
if $programname == 'arcmgrd'            then :omfile:$arcmgrd_log;exaLog
if $programname == 'audmgrd'            then :omfile:$audmgrd_log;exaLog
if $programname == 'avonmgrd'           then :omfile:$avonmgrd_log;exaLog
if $programname == 'bngaxd'             then :omfile:$bngaxd_log;exaLog
if $programname == 'bngaxlid'           then :omfile:$bngaxlid_log;exaLog
if $programname == 'bnglcmd'            then :omfile:$bnglcmd_log;exaLog
#if $programname == 'bnglid'             then :omfile:$bnglid_log;exaLog
if $programname == 'certmgrd'           then :omfile:$certmgrd_log;exaLog
if $programname == 'chassismgrd'        then :omfile:$chassismgrd_log;exaLog
if $programname == 'confd'              then :omfile:$confd_log;thdptyLog
if $programname == 'cpeupgmgrd'         then :omfile:$cpeupgmgrd_log;exaLog
if $programname == 'dadd'               then :omfile:$dadd_log;exaLog
if $programname == 'dgtld'              then :omfile:$dgtld_log;exaLog
if $programname == 'dhcpd'              then :omfile:$dhcpd_log;thdptyLog
if $programname == 'dhclient'           then :omfile:$dhclient_log;thdptyLog
if $programname == 'diagmgrd'           then :omfile:$diagmgrd_log;exaLog
if $programname == 'diagtestd'          then :omfile:$diagtestd_log;exaLog
if $programname == 'diameterd'          then :omfile:$diameterd_log;exaLog
if $programname == 'diametergyd'        then :omfile:$diametergyd_log;exaLog
if $programname == 'dmd'                then :omfile:$dmd_log;exaLog
if $programname == 'dpmgrd'             then :omfile:$dpmgrd_log;exaLog
if $programname == 'dyadmgrd'           then :omfile:$dyadmgrd_log;exaLog
if $programname == 'eapsd'              then :omfile:$eapsd_log;exaLog
if $programname == 'equipmentmgrd'      then :omfile:$equipmentmgrd_log;exaLog
if $programname == 'erpsd'              then :omfile:$erpsd_log;exaLog
if $programname == 'evtest'             then :omfile:$evtest_log;exaLog
if $programname == 'fald'               then :omfile:$fald_log;exaLog
if $programname == 'fccd'               then :omfile:$fccd_log;exaLog
if $programname == 'fibd'               then :omfile:$fibd_log;exaLog
if $programname == 'flwmgrd'            then :omfile:$flwmgrd_log;exaLog
if $programname == 'fprd'               then :omfile:$fprd_log;exaLog
if $programname == 'gfastmgrd'          then :omfile:$gfastmgrd_log;exaLog
if $programname == 'gfastdslmgrd'       then :omfile:$gfastdslmgrd_log;exaLog
if $programname == 'halm_bcmfastdsl'    then :omfile:$halm_bcmfastdsl_log;exaLog
if $programname == 'halm_bcmxgsu'       then :omfile:$halm_bcmxgsu_log;exaLog
if $programname == 'halm_xgs'           then :omfile:$halm_xgs_log;exaLog
if $programname == 'halm_dyaddev'       then :omfile:$halm_dyaddev_log;exaLog
if $programname == 'halm_np'            then :omfile:$halm_np_log;exaLog
if $programname == 'halm_dnx'           then :omfile:$halm_dnx_log;exaLog
if $programname == 'halm_sim'           then :omfile:$halm_sim_log;exaLog
if $programname == 'halm_ovs'           then :omfile:$halm_ovs_log;exaLog
if $programname == 'halm_sckipio'       then :omfile:$halm_sckipio_log;exaLog
if $programname == 'halm_robo'          then :omfile:$halm_robo_log;exaLog
if $programname == 'halm_vmac'          then :omfile:$halm_vmac_log;exaLog
if $programname == 'halm_vswitch'       then :omfile:$halm_vswitch_log;exaLog
if $programname == 'halm_vitesse'       then :omfile:$halm_vitesse_log;exaLog
if $programname == 'haltestd'           then :omfile:$haltestd_log;exaLog
if $programname == 'hatestd'            then :omfile:$hatestd_log;exaLog
if $programname == 'hostapd'            then :omfile:$hostapd_log;thdptyLog
if $programname == 'hostmgrd'           then :omfile:$hostmgrd_log;exaLog
if $programname == 'httprd'             then :omfile:$httprd_log;exaLog
if $programname == 'icld'               then :omfile:$icld_log;exaLog
if $programname == 'ictpd'              then :omfile:$ictpd_log;exaLog
if $programname == 'ifmgrd'             then :omfile:$ifmgrd_log;exaLog
if $programname == 'ifmd'               then :omfile:$ifmd_log;exaLog
if $programname == 'igmpd'              then :omfile:$igmpd_log;exaLog
if $programname == 'ipmgrd'             then :omfile:$ipmgrd_log;exaLog
if $programname == 'ipcmgrd'            then :omfile:$ipcmgrd_log;exaLog
if $programname == 'ipdrd'              then :omfile:$ipdrd_log;exaLog
if $programname == 'ipfixd'             then :omfile:$ipfixd_log;exaLog
if $programname == 'l2dd'               then :omfile:$l2dd_log;exaLog
if $programname == 'l2hostmgrd'         then :omfile:$l2hostmgrd_log;exaLog
if $programname == 'l2mgrd'             then :omfile:$l2mgrd_log;exaLog
if $programname == 'l2mmgrd'            then :omfile:$l2mmgrd_log;exaLog
if $programname == 'lacpd'              then :omfile:$lacpd_log;exaLog
if $programname == 'lldpd'              then :omfile:$lldpd_log;exaLog
if $programname == 'pppsrvd'            then :omfile:$pppsrvd_log;exaLog
if $programname == 'pppsrvdp_proxyd'   then :omfile:$pppsrvdp_proxyd_log;exaLog
if $programname == 'vrrpd'              then :omfile:$vrrpd_log;exaLog
if $programname == 'keepalived'         then :omfile:$keepalived_log;exaLog

#use the AT- tag to differentiate Audit log from regular lmd messages
if $programname == 'lmd' and not($msg contains 'AT-') then :omfile:$lmd_log;exaLog
if $programname == 'lmd' and ($msg contains 'AT-') then :omfile:$cmd_log;exaLog

if $programname == 'loamd'              then :omfile:$loamd_log;exaLog
if $programname == 'login'              then :omfile:$login_log;exaLog
if $programname == 'logmgrd'            then :omfile:$logmgrd_log;exaLog
if $programname == 'macsecd'            then :omfile:$macsecd_log;exaLog
if $programname == 'notfmgrd'           then :omfile:$notfmgrd_log;exaLog
if $programname == 'notfproxyd'         then :omfile:$notfproxyd_log;exaLog
if $programname == 'ntpd'               then :omfile:$ntpd_log;thdptyLog
if $programname == 'ofmgrd'             then :omfile:$ofmgrd_log;exaLog
if $programname == 'olttcmgrd'          then :omfile:$olttcmgrd_log;exaLog
if $programname == 'ontimgcd'           then :omfile:$ontimgcd_log;exaLog
if $programname == 'ontimgfd'           then :omfile:$ontimgfd_log;exaLog
if $programname == 'ontupgmd'           then :omfile:$ontupgmd_log;exaLog
if $programname == 'ontupgsd'           then :omfile:$ontupgsd_log;exaLog
if $programname == 'opentftpd'          then :omfile:$opentftpd_log;thdptyLog
if $programname == 'pcmd'               then :omfile:$pcmd_log;exaLog
if $programname == 'plmgrd'             then :omfile:$plmgrd_log;exaLog
if $programname == 'ponmgrd'            then :omfile:$ponmgrd_log;exaLog
if $programname == 'pppoeiad'           then :omfile:$pppoeiad_log;exaLog
if $programname == 'qosmgrd'            then :omfile:$qosmgrd_log;exaLog
if $programname == 'radiusd'            then :omfile:$radiusd_log;exaLog
if $programname == 'rstpd'              then :omfile:$rstpd_log;exaLog
if $programname == 'rsyncd'             then :omfile:$rsyncd_log;thdptyLog
# In case of rate-limiting, programname is rsyslogd-<pid>, therefore we use 'startswith' here:
if $programname startswith 'rsyslogd'   then :omfile:$rsyslogd_log;thdptyLog
if $programname == 'satmgrd'            then :omfile:$satmgrd_log;exaLog
if $programname == 'servicemgrd'        then :omfile:$servicemgrd_log;exaLog
if $programname == 'snmpd'              then :omfile:$snmpd_log;exaLog
if $programname == 'soamd'              then :omfile:$soamd_log;exaLog
if $programname == 'ssal_mgrd'          then :omfile:$ssal_mgrd_log;exaLog
if $programname == 'ssaliod'            then :omfile:$ssaliod_log;exaLog
if $programname == 'ssalio2d'           then :omfile:$ssalio2d_log;exaLog
if $programname == 'ssmd'               then :omfile:$ssmd_log;exaLog
if $programname == 'sshd'               then :omfile:$sshd_log;exaLog
if $programname == 'tmmgrd'             then :omfile:$tmmgrd_log;exaLog
if $programname == 'tracemgrd'          then :omfile:$tracemgrd_log;exaLog
if $programname == 'uamgrd'             then :omfile:$uamgrd_log;exaLog
if $programname == 'upgmgrd'            then :omfile:$upgmgrd_log;exaLog
if $programname == 'usbmgrd'            then :omfile:$usbmgrd_log;exaLog
if $programname == 'vcamgrd'            then :omfile:$vcamgrd_log;exaLog
if $programname == 'vlanmgrd'           then :omfile:$vlanmgrd_log;exaLog
if $programname == 'l2v6hostmgrd'       then :omfile:$l2v6hostmgrd_log;exaLog
if $programname == 'salusd'             then :omfile:$salusd_log;exaLog

template(name="bngliLogLine" type="string" string="%timegenerated:::date-rfc3339% [%syslogpriority%] %msg:F,44:2%\n")
template(name="bngliLog" type="string" string="/var/lilog/%$!user%/%$!dyn_fname%")
$outchannel bnglilea_log, /var/lilog/lea/lea.log,   500000, /usr/sbin/logrotate /etc/logrotate.conf

if ($programname == "bnglid") then {
  #extract username from log msg, msg is formatted as (user=liuser1 msg content)
  set $!v1 = re_extract($msg, "user@([a-z0-9])+", 0, 0,"user@!");
  set $!user = field($!v1, 64, 2);
  #drop msgs that doesn't have user specified
  if ($!user == "!") then stop
  #set dynamic file name
  set $!dyn_fname = $!user & ".log";
  #output file
  if ($!user != "lea") then {
    action(type="omfile" dynaFile="bngliLog" template="bngliLogLine")
  }
  :omfile:$bnglilea_log;bngliLogLine
  stop
}

# All important logs go to imessages
daemon.warn     :omfile:$important_log;thdptyLog
local0.warn     :omfile:$important_log;exaLog
kern.warn       :omfile:$important_log;thdptyLog

# Important kernel messages also go to /var/log/kernel
kern.notice     :omfile:$kernel_log;thdptyLog

# These 3rd party daemons have their own logs above, so they don't need to go to /var/log/messages:
:programname, isequal,      "confd"     stop
:programname, isequal,      "dhcpd"     stop
:programname, isequal,      "keepalived" stop
:programname, isequal,      "dhclient"  stop
:programname, isequal,      "hostapd"   stop
:programname, isequal,      "ntpd"      stop
:programname, isequal,      "rsyncd"    stop
:programname, isequal,      "opentftpd" stop
:programname, startswith,   "rsyslogd"  stop

# Catch the rest of the daemons not covered above
daemon.*        :omfile:$system_log;thdptyLog

# logging host configuration section provided by log-mgr: loghosts.conf
$IncludeConfig /etc/rsyslog.d/*.conf
`

processImInternal  rsyslogd.c Line 1655
    --- logmsgInternal_doWrite() --- Line 1666
          --- submitMsg2(pMsg) --- Line 851
                --- qqueueEnqMsg()  --- Line 1082
                      --- CHKiRet(doEnqSingleObj(pThis, flowCtlType, pMsg)); --- Line 3132 runtime/queue.c

gdb backtrace of rsyslogd when hung problem happened, almost hung for 40- 80 seconds, that delay the application which call vsyslog() from glibc.  Not sure restart rsyslog will solve it.
`Thread 8 (LWP 14209):
#0  0x0fd708a8 in pthread_cond_timedwait () from /lib/libpthread.so.0
No symbol table info available.
#1  0x1005e734 in ?? ()
No symbol table info available.
#2  0x10062290 in qqueueEnqMsg ()
No symbol table info available.
#3  0x1006f418 in ?? ()
No symbol table info available.
#4  0x1006593c in ?? ()
No symbol table info available.
#5  0x10065910 in ?? ()
No symbol table info available.
#6  0x100661c8 in ?? ()
No symbol table info available.
#7  0x1000b7c4 in ?? ()
No symbol table info available.
#8  0x10060cc8 in ?? ()
No symbol table info available.
#9  0x1005cd24 in wtiWorker ()
No symbol table info available.
#10 0x1005ac64 in ?? ()
No symbol table info available.
#11 0x0fd67514 in ?? () from /lib/libpthread.so.0
No symbol table info available.
#12 0x0feecb60 in clone () from /lib/libc.so.6
No symbol table info available.

Thread 7 (LWP 14083):
#0  0x0fee3a9c in select () from /lib/libc.so.6
No symbol table info available.
#1  0x1004a360 in srSleep ()
No symbol table info available.
#2  0x1006c81c in ?? ()
No symbol table info available.
#3  0x1006e0dc in ?? ()
No symbol table info available.
#4  0x1006f2d4 in ?? ()
No symbol table info available.
#5  0x10060cc8 in ?? ()
No symbol table info available.
#6  0x1005cd24 in wtiWorker ()
No symbol table info available.
#7  0x1005ac64 in ?? ()
No symbol table info available.
#8  0x0fd67514 in ?? () from /lib/libpthread.so.0
No symbol table info available.
#9  0x0feecb60 in clone () from /lib/libc.so.6
No symbol table info available.

Thread 6 (LWP 14004):
#0  0x0fd708a8 in pthread_cond_timedwait () from /lib/libpthread.so.0
No symbol table info available.
#1  0x1005e734 in ?? ()
No symbol table info available.
#2  0x10062290 in qqueueEnqMsg ()
No symbol table info available.
#3  0x1006f418 in ?? ()
No symbol table info available.
#4  0x1006593c in ?? ()
No symbol table info available.
#5  0x10065910 in ?? ()
No symbol table info available.
#6  0x100661c8 in ?? ()
No symbol table info available.
#7  0x1000b7c4 in ?? ()
No symbol table info available.
#8  0x10060cc8 in ?? ()
No symbol table info available.
#9  0x1005cd24 in wtiWorker ()
No symbol table info available.
#10 0x1005ac64 in ?? ()
No symbol table info available.
#11 0x0fd67514 in ?? () from /lib/libpthread.so.0
No symbol table info available.
#12 0x0feecb60 in clone () from /lib/libc.so.6
No symbol table info available.

Thread 5 (LWP 14002):
#0  0x0feed270 in epoll_wait () from /lib/libc.so.6
No symbol table info available.
#1  0x0f195744 in ?? () from /usr/lib/rsyslog/imudp.so
No symbol table info available.
#2  0x0f196254 in ?? () from /usr/lib/rsyslog/imudp.so
No symbol table info available.
#3  0x0f19638c in ?? () from /usr/lib/rsyslog/imudp.so
No symbol table info available.
#4  0x1007150c in ?? ()
No symbol table info available.
#5  0x0fd67514 in ?? () from /lib/libpthread.so.0
No symbol table info available.
#6  0x0feecb60 in clone () from /lib/libc.so.6
No symbol table info available.

Thread 4 (LWP 14001):
#0  0x0fee3a9c in select () from /lib/libc.so.6
No symbol table info available.
#1  0x1004a360 in srSleep ()
No symbol table info available.
#2  0x0f1f781c in ?? () from /usr/lib/rsyslog/impstats.so
No symbol table info available.
#3  0x1007150c in ?? ()
No symbol table info available.
#4  0x0fd67514 in ?? () from /lib/libpthread.so.0
No symbol table info available.
#5  0x0feecb60 in clone () from /lib/libc.so.6
No symbol table info available.

Thread 3 (LWP 14000):
#0  0x0fd76564 in read () from /lib/libpthread.so.0
No symbol table info available.
#1  0x0f228c3c in ?? () from /usr/lib/rsyslog/imklog.so
No symbol table info available.
#2  0x0f228fe4 in klogLogKMsg () from /usr/lib/rsyslog/imklog.so
No symbol table info available.
#3  0x0f227f2c in ?? () from /usr/lib/rsyslog/imklog.so
No symbol table info available.
#4  0x1007150c in ?? ()
No symbol table info available.
#5  0x0fd67514 in ?? () from /lib/libpthread.so.0
No symbol table info available.
#6  0x0feecb60 in clone () from /lib/libc.so.6
No symbol table info available.

Thread 2 (LWP 13999):
#0  0x0fd708a8 in pthread_cond_timedwait () from /lib/libpthread.so.0
No symbol table info available.
#1  0x1005e734 in ?? ()
No symbol table info available.
#2  0x10062290 in qqueueEnqMsg ()
No symbol table info available.
#3  0x1000db1c in submitMsg2 ()
No symbol table info available.
#4  0x10067918 in ratelimitAddMsg ()
No symbol table info available.
#5  0x0f25b580 in ?? () from /usr/lib/rsyslog/imuxsock.so
No symbol table info available.
#6  0x0f25c35c in ?? () from /usr/lib/rsyslog/imuxsock.so
No symbol table info available.
#7  0x1007150c in ?? ()
No symbol table info available.
#8  0x0fd67514 in ?? () from /lib/libpthread.so.0
No symbol table info available.
#9  0x0feecb60 in clone () from /lib/libc.so.6
No symbol table info available.

Thread 1 (LWP 13998):
#0  0x0fd708a8 in pthread_cond_timedwait () from /lib/libpthread.so.0
No symbol table info available.
#1  0x1005e734 in ?? ()
No symbol table info available.
#2  0x10062290 in qqueueEnqMsg ()
No symbol table info available.
#3  0x1000db1c in submitMsg2 ()
No symbol table info available.
#4  0x1000de14 in ?? ()
No symbol table info available.
#5  0x1000dfd8 in processImInternal ()
No symbol table info available.
#6  0x1000aed4 in main ()
No symbol table info available.`

loghosts.conf, note: 192.168.5.114 is the problem setup ip, 192.168.5.17 is the remote rsyslogd server
`# Provides UDP syslog reception
# for parameters see http://www.rsyslog.com/doc/imudp.html
# DO NOT REMOVE below commented lines and the comment at the end
module(load="imudp") # udp Listening - Needs to be done only once.
# input(type="imudp" port="514")  # udp eth2 Listening
input(type="imudp" port="514" address="192.168.5.114")   # udp net2 Listening

# Provides TCP syslog reception
# for parameters see http://www.rsyslog.com/doc/imtcp.html
# DO NOT REMOVE below commented lines and the  comment at the end
# module(load="imtcp") # tcp Listening - Needs to be done only once.
# input(type="imtcp" port="514")  # tcp eth2 Listening 
# input(type="imtcp" port="514")  # tcp net2 Listening

### Do not remove the below two commented lines. ###
if ( ($fromhost-ip != '192.168.5.114') and ($fromhost-ip != '127.0.0.1') ) then stop # FromHost-IP

#$WorkDirectory /var/lib/rsyslog # where to place spool files
#$ActionQueueFileName fwdRule1 # unique name prefix for spool files
#$ActionQueueMaxDiskSpace 1g   # 1gb space limit (use as much as possible)
#$ActionQueueSaveOnShutdown on # save messages to disk on shutdown
#$ActionQueueType LinkedList   # run asynchronously
#$ActionResumeRetryCount -1    # infinite retries if host is down
#$ActionQueueHighWaterMark 8000
#$ActionQueueFileName papertrailqueue
#$ActionQueueMaxDiskSpace 2g
#$ActionQueueSaveOnShutdown on
# remote host is: name/ip:port, e.g. 192.168.0.1:514, port optional

#
# Internal Log forwarding.
# Internal hosts : 1
#
# start internal forwarding rule 1
$ActionResumeInterval 10
$ActionQueueSize 10000
$ActionQueueDiscardMark 9750
$ActionQueueType LinkedList
$ActionQueueCheckpointInterval 100
$ActionResumeRetryCount 10
$ActionQueueTimeoutEnqueue 100
$ActionQueueDiscardSeverity 0
*.warn;kern.none action(name="send_remote" type="omfwd" Target="192.168.5.17" Port="514" Protocol="UDP" queue.type="LinkedList" queue.size="50000" queue.timeoutEnqueue="800" action.resumeRetryCount="-1" action.resumeInterval="2")  # InTernaL HoSt`

**Comment by fduan001** [comment]
application's backtrace indicate it hung as below. Our application heartbeat in period, if HB out of 30 seconds, it will dump the stack of application and rsyslogd
  224 Thread 1 (LWP 6249):
  225 #0  0x0feeee90 in send () from /lib/libc.so.6
  226 #1  0x0fee77d4 in __vsyslog_chk () from /lib/libc.so.6
  227 #2  0x0f7e4d18 in dl_vlog_print_facility () from /usr/lib/libdaemonlib.so.3
  228 #3  0x0f7e5018 in dl_vlog_print () from /usr/lib/libdaemonlib.so.3
  229 #4  0x0f7e50bc in dl_log_print_tag () from /usr/lib/libdaemonlib.so.3
  230 #5  0x100c34a4 in ifm_evt_cb(void*) ()
  231 #6  0x0f7df078 in event_fd_callback () from /usr/lib/libdaemonlib.so.3
  232 #7  0x0f7e8bd4 in dl_poll_fds_ctx () from /usr/lib/libdaemonlib.so.3
  233 #8  0x0f7e8d64 in dl_poll_fds () from /usr/lib/libdaemonlib.so.3
  234 #9  0x10024fe8 in main ()

**Comment by dongbiao830415** [comment]
我的程序在使用vsyslog发送日志时，也遇到了这个问题，其它使用vsyslog的进程也都卡住了，重启rsyslog双恢复正常了，这个问题你是如何解决的?谢谢！

**Comment by rgerhards** [comment]
Sorry for the very long delay. rsyslog does not lightly close issues and I am not closing this just because it is old; I checked the retained evidence while working through the backlog.

The actionable evidence here is not enough to tie this to a current rsyslog bug. The supplied stack trace is from an application blocked in libc `send()` via `__vsyslog_chk()`, not an rsyslogd backtrace or a reproducible rsyslog-side failure path. The original environment was also rsyslog 1904 on Yocto and the report says it was hard to reproduce.

That application-side blocking can happen when the local syslog socket path applies backpressure or the receiver is not consuming, but this issue does not retain the rsyslogd debug log/backtrace, current-version config, or reproduction details needed to fix something in rsyslog. I am closing this as not reproducible / needs current reporter info. If this still happens on a current rsyslog version, please open a new issue with the rsyslog version, full config, rsyslogd thread backtraces/debug log while stuck, and the local socket details.
